kernel/constr.ml

```diff
@@ -131,78 +130,103 @@ type rec_declaration = (constr, types) prec_declaration
 type fixpoint = (constr, types) pfixpoint
 type cofixpoint = (constr, types) pcofixpoint
 
+(************************************************************************)
+(*    kind_of_term = constructions as seen by the user                 *)
+(************************************************************************)
+
+(* User view of [constr]. For [App], it is ensured there is at
+   least one argument and the function is not itself an applicative
+   term *)
+
+let kind (T c) = c
+
+let rec kind_nocast_gen kind c =
+  match kind c with
+  | Cast (c, _, _) -> kind_nocast_gen kind c
+  | App (h, outer) as k ->
+    (match kind_nocast_gen kind h with
+     | App (h, inner) -> App (h, Array.append inner outer)
+     | _ -> k)
+  | k -> k
+
+let kind_nocast c = kind_nocast_gen kind c
+
 (*********************)
 (* Term constructors *)
 (*********************)
 
 (* Constructs a de Bruijn index with number n *)
-let rels =
-  [|Rel  1;Rel  2;Rel  3;Rel  4;Rel  5;Rel  6;Rel  7; Rel  8;
-    Rel  9;Rel 10;Rel 11;Rel 12;Rel 13;Rel 14;Rel 15; Rel 16|]
+let rels = Array.init 17 (fun i -> T (Rel i))
```

Comment:
This is now missing a + 1.

Reply:
(I mean that we don't want to cache Rel 0)

Reply:
Why not? It's simpler this way IMO

Reply:
I find it weird and potentially error-inducing, but if you don't really care I'm also fine with it as-is.